src/renderer/atlas/AtlasEngine.h

```diff
@@ -0,0 +1,761 @@
+                if constexpr (Alignment <= __STDCPP_DEFAULT_NEW_ALIGNMENT__)
+                {
+                    ::operator delete(data);
+                }
+                else
+                {
+                    ::operator delete(data, static_cast<std::align_val_t>(Alignment));
+                }
+            }
+#pragma warning(pop)
+
+            T* _data = nullptr;
+            size_t _size = 0;
+        };
+
+        // This structure works similar to how std::string works:
+        // You can think of a std::string as a structure consisting of:
+        //   char*  data;
+        //   size_t size;
+        //   size_t capacity;
+        // where data is some backing memory allocated on the heap.
+        //
+        // But std::string employs an optimization called "small string optimization" (SSO).
+        // To simplify things it could be explained as:
+        // If the string capacity is small, then the characters are stored inside the "data"
+        // pointer and you make sure to set the lowest bit in the pointer one way or another.
+        // Heap allocations are always aligned by at least 4-8 bytes on any platform.
+        // If the address of the "data" pointer is not even you know data is stored inline.
+        template<typename T>
+        union SmallObjectOptimizer
+        {
+            static_assert(std::is_trivially_copyable_v<T>);
+            static_assert(std::has_unique_object_representations_v<T>);
+
+            T* allocated = nullptr;
+            T inlined;
+
+            constexpr SmallObjectOptimizer() = default;
+
+            SmallObjectOptimizer(const SmallObjectOptimizer& other)
+            {
+                const auto otherData = other.data();
+                const auto otherSize = other.size();
+                const auto data = initialize(otherSize);
+                memcpy(data, otherData, otherSize);
+            }
+
+            SmallObjectOptimizer& operator=(const SmallObjectOptimizer& other)
+            {
+                if (this != &other)
+                {
+                    delete this;
+                    new (this) SmallObjectOptimizer(other);
+                }
+                return &this;
+            }
+
+            SmallObjectOptimizer(SmallObjectOptimizer&& other) noexcept
+            {
+                memcpy(this, &other, std::max(sizeof(allocated), sizeof(inlined)));
+                other.allocated = nullptr;
+            }
+
+            SmallObjectOptimizer& operator=(SmallObjectOptimizer&& other) noexcept
+            {
+                return *new (this) SmallObjectOptimizer(other);
+            }
+
+            ~SmallObjectOptimizer()
+            {
+                if (!is_inline())
+                {
+#pragma warning(suppress : 26408) // Avoid malloc() and free(), prefer the nothrow version of new with delete (r.10).
+                    free(allocated);
+                }
+            }
+
+            T* initialize(size_t byteSize)
+            {

```

Comment:
I believe it's UB to not placement-new a `T` into the returned space, even if T `is_trivially_copyable` and `has_unique_object_representation`.

Reply:
Oh it totally is! I can't even use `malloc()` here, because `malloc()` is UB in C++ in general. P0593R6 should deal with this.
Until then we should just assume that the compiler will be reasonable and treat neither `malloc()` nor `operator new()` as UB....